addons/account/models/account_payment_term.py

```diff
@@ -95,6 +124,62 @@ def unlink(self):
             ).unlink()
         return super(AccountPaymentTerm, self).unlink()
 
+    @api.depends('line_ids')
+    def _compute_early_payment_applicable(self):
+        for term in self:
+            term.early_payment_applicable = not any(line.value != 'balance' for line in term.line_ids)
+
+    @api.depends('early_payment_applicable')
+    def _compute_has_early_payment(self):
+        for term in self:
+            if not term.early_payment_applicable:
+                term.has_early_payment = False
+
+    @api.onchange('discount_days')
+    def _onchange_discount_days(self):
+        if self.discount_days <= 0:
+            raise ValidationError(_("The discount availability must be strictly positive."))
+        if self.discount_time_availability == "day_following_month" or self.discount_time_availability == "day_current_month":
+            if self.discount_days > 31:
+                raise ValidationError(_("This date is not valid."))
+
+    @api.onchange('discount_time_availability')
+    def _onchange_discount_time_availability(self):
+        if self.discount_time_availability == "day_following_month" or self.discount_time_availability == "day_current_month":
+            if self.discount_days > 31:
+                raise ValidationError(_("This date is not valid."))
+
+    @api.onchange('percentage_to_discount')
+    def _onchange_percentage_to_discount(self):
+        if self.percentage_to_discount <= 0:
+            raise ValidationError(_("The discount percentage must be strictly positive."))
+        if self.percentage_to_discount > 100:
+            raise ValidationError(_("The discount percentage cannot exceed 100."))
+
+    def _get_last_date_for_discount(self, move_date):
+        if self.discount_time_availability == 'days_after_invoice_date':
+            return move_date + timedelta(days=self.discount_days)
+        if self.discount_time_availability == 'days_after_invoice_month':
+            return date_utils.end_of(move_date, 'month') + timedelta(self.discount_days)
+        if self.discount_time_availability == 'day_following_month':
+            computed_date = date_utils.end_of(move_date, 'month') + timedelta(self.discount_days)
+            # Check that the number of days in the timedelta arent > to the number of days in the month
+            # (ex: February 31st). If it happens, return the last day of the next month
+            last_day_of_month = date_utils.get_month(computed_date)[1]
+            if last_day_of_month < computed_date:
+                return last_day_of_month
+            return computed_date
+        if self.discount_time_availability == 'day_current_month':
+            computed_date = date_utils.start_of(fields.Date.context_today(self), 'month') + timedelta(self.discount_days)
```

Comment:
If you set `self.discount_days = 20`, you are still considered as in the EPD the 21th. So I think you need to remove one day
Edit : or
```py
today = fields.Date.context_today(self)
last_day_of_month = date_utils.get_month(computed_date)[1]
computed_date = today.replace(day=min(last_day_of_month, self.discount_days))
```